relay/sources/relayd/src/api/metrics.rs

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// SPDX-FileCopyrightText: 2019-2020 Normation SAS
+
+use crate::metrics::REGISTRY;
+use warp::{filters::method, path, Filter, Reply};
+
+/// Special case for /metrics, standard for prometheus
+pub fn routes() -> impl Filter<Extract = impl Reply, Error = warp::Rejection> + Clone {
+    method::get()
+        .and(path!("metrics"))
+        .and_then(handlers::metrics)
+}
+
+pub mod handlers {
+    use super::*;
+    use crate::api::RudderReject;
+    use warp::{reject, Rejection, Reply};
+
+    pub async fn metrics() -> Result<impl Reply, Rejection> {
+        use prometheus::Encoder;
+        let encoder = prometheus::TextEncoder::new();
+
+        let mut buffer = Vec::new();
+        if let Err(e) = encoder.encode(&REGISTRY.gather(), &mut buffer) {
+            return Err(reject::custom(RudderReject::new(e)));
+        };
+        let mut res = match String::from_utf8(buffer.clone()) {
+            Ok(v) => v,
+            Err(e) => return Err(reject::custom(RudderReject::new(e))),
+        };
+        buffer.clear();
+
+        let mut buffer = Vec::new();
```

Comment:
since it has just been cleared I guess you dont need to declare buffer again

Reply:
You could also define a closure (called twice) to do the encoding

Reply:
I have to admit I just copy-pasted the lib example here, but it can be factored easily

Reply:
A closure is indeed way better!

```rust
    pub async fn metrics() -> Result<impl Reply, Rejection> {
        use prometheus::Encoder;
        let encoder = prometheus::TextEncoder::new();
        let mut buffer = Vec::new();
        let mut encode = |metrics: &[MetricFamily]| match encoder.encode(&metrics, &mut buffer) {
            Ok(_) => Ok(()),
            Err(e) => Err(reject::custom(RudderReject::new(e))),
        };

        encode(&REGISTRY.gather())?;
        encode(&prometheus::gather())?;
        Ok(buffer)
    }
```